README.md

# Flutter iOS Ringtone Player

The purpose of this plugin is to play iOS system sound.


## Authors

- [@hadrienm](https://www.github.com/hadrienm)


## Installation

```bash
  flutter pub add flutter_ios_ringtone_player
```
    
## Usage/Examples

```dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();

    initPlatformState();
    play();
  }

  Future<void> initPlatformState() async {
    String platformVersion;
    try {
      platformVersion =
          await FlutterIosRingtonePlayer.getPlatformVersion() ?? 'Unknown platform version';
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  Future<void> play() async {
    FlutterIosRingtonePlayer.play(soundId: 1035);
    Future.delayed(const Duration(milliseconds: 500), () async {
       FlutterIosRingtonePlayer.stop(soundId: 1035);
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_platformVersion\n'),
        ),
      ),
    );
  }
}
```


## Features

- Play iOS system sound


## Documentation

play() method:

    Play a given sound.

    Take soundId as unique parameter, is must be a int. It must be between 1,000 and 4,000

stop() method:

    Stop the current playing sound. Take soundId as unique parameter, is must be a int. It must be between 1,000 and 4,000



CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* Indicates the platforms available on pub.dev

## 0.0.3

* Update dependencies, and improve the description of the plugin

## 0.0.4

* change sdk version

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_ios_ringtone_player_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_ios_ringtone_player_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player.dart';
import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player_platform_interface.dart';
import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player_method_channel.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

class MockFlutterIosRingtonePlayerPlatform
    with MockPlatformInterfaceMixin
    implements FlutterIosRingtonePlayerPlatform {

  @override
  Future<String?> getPlatformVersion() => Future.value('42');
}

void main() {
  final FlutterIosRingtonePlayerPlatform initialPlatform = FlutterIosRingtonePlayerPlatform.instance;

  test('$MethodChannelFlutterIosRingtonePlayer is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelFlutterIosRingtonePlayer>());
  });

  test('getPlatformVersion', () async {
    FlutterIosRingtonePlayer flutterIosRingtonePlayerPlugin = FlutterIosRingtonePlayer();
    MockFlutterIosRingtonePlayerPlatform fakePlatform = MockFlutterIosRingtonePlayerPlatform();
    FlutterIosRingtonePlayerPlatform.instance = fakePlatform;

    expect(await flutterIosRingtonePlayerPlugin.getPlatformVersion(), '42');
  });
}


example/integration_test/plugin_integration_test.dart
// This is a basic Flutter integration test.
//
// Since integration tests run in a full Flutter application, they can interact
// with the host side of a plugin implementation, unlike Dart unit tests.
//
// For more information about Flutter integration tests, please see
// https://docs.flutter.dev/cookbook/testing/integration/introduction


import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';

import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player.dart';

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('getPlatformVersion test', (WidgetTester tester) async {
    final FlutterIosRingtonePlayer plugin = FlutterIosRingtonePlayer();
    final String? version = await plugin.getPlatformVersion();
    // The version string depends on the host platform running the test, so
    // just assert that some non-empty string is returned.
    expect(version?.isNotEmpty, true);
  });
}


ios/Classes/FlutterIosRingtonePlayerPlugin.swift
import Flutter
import UIKit
import AVFoundation


class SoundPlayer {
    static var shared = SoundPlayer()

    private var audioSession: AVAudioSession

    init() {
        audioSession = AVAudioSession.sharedInstance()
        setupAudioSession()
    }

    private func setupAudioSession() {
        do {
            try audioSession.setCategory(.playback, mode: .default, options: [])
            try audioSession.setActive(true)
        } catch {
            print("Error setting up AVAudioSession: \(error)")
        }
    }

    func playSystemSound(soundId: SystemSoundID) {
        AudioServicesPlaySystemSound(soundId)
    }

    func vibrate() {
        AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
    }

    func stopSound(soundId: SystemSoundID) {
        AudioServicesDisposeSystemSoundID(soundId)
    }
}

public class FlutterIosRingtonePlayerPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_ios_ringtone_player", binaryMessenger: registrar.messenger())
    let instance = FlutterIosRingtonePlayerPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "play":
        let arguments = call.arguments as! [String: Any]
        let soundId = arguments["soundId"] as! UInt32

        SoundPlayer.shared.playSystemSound(soundId: soundId)
    case "stop":
        let arguments = call.arguments as! [String: Any]
        let soundId = arguments["soundId"] as! UInt32

        SoundPlayer.shared.stopSound(soundId: soundId)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_ios_ringtone_player/flutter_ios_ringtone_player.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();

    initPlatformState();
    play();
  }

  Future<void> initPlatformState() async {
    String platformVersion;
    try {
      platformVersion =
          await FlutterIosRingtonePlayer.getPlatformVersion() ?? 'Unknown platform version';
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  Future<void> play() async {
    FlutterIosRingtonePlayer.play(soundId: 1035);
    Future.delayed(const Duration(milliseconds: 500), () async {
       FlutterIosRingtonePlayer.stop(soundId: 1035);
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on: $_platformVersion\n'),
        ),
      ),
    );
  }
}
